A64 fastmem

vdwjeremy opened the pull request:
Direct access to host (virtual) address space with ```fastmem``` offset, propagated from A32.
The goal is to delegate page table computation and page boundaries detection to the hardware.

**Benchmark**
From memory intensive sequential read/write
```
    env.MemoryWrite32(0, 0x910004E7); // ADD X7, X7, 1
    env.MemoryWrite32(4, 0x91002042); // ADD X2, X2, 8
    env.MemoryWrite32(8, 0xF9400046); // LDR X6, [X2]
    env.MemoryWrite32(12, 0x91002063); // ADD X3, X3, 8
    env.MemoryWrite32(16, 0xF9000066); // STR X6, [X3]
    env.MemoryWrite32(20, 0xF9400046); // LDR X6, [X2]
    env.MemoryWrite32(24, 0xF9000066); // STR X6, [X3]
    env.MemoryWrite32(28, 0xF9400046); // LDR X6, [X2]
    env.MemoryWrite32(32, 0xF9000066); // STR X6, [X3]
    env.MemoryWrite32(36, 0xF9400046); // LDR X6, [X2]
    env.MemoryWrite32(40, 0xF9000066); // STR X6, [X3]
    env.MemoryWrite32(44, 0x17FFFFF5); // B -24
```
Elapse time on i5-4590 compared to page table is

- 70% when copying 2MB
- 71% when copying 20MB
- 78% when copying 200MB

lioncash commented on src/backend/x64/a64_emit_x64.cpp:
```diff
@@ -877,51 +934,76 @@ void A64EmitX64::EmitDirectPageTableMemoryWrite(A64EmitContext& ctx, IR::Inst* i
+    if (conf.fastmem_pointer) {
+        EmitDirectMemoryRead(ctx, inst, 8);
+    }
+    else if (conf.page_table) {
```
Our formatting combines the braces onto one line like so:

```suggestion
    } else if (conf.page_table) {
```

ditto for all other occurrences

lioncash commented on tests/A64/testenv.h:
```diff
@@ -125,3 +125,89 @@ class A64TestEnv final : public Dynarmic::A64::UserCallbacks {
+        return *((uint64_t*)backing_memory + (vaddr / 8));
+    }
+    Vector MemoryRead128(u64 vaddr) override {
+        return { MemoryRead64(vaddr), MemoryRead64(vaddr + 8) };
```
```suggestion
        return {MemoryRead64(vaddr), MemoryRead64(vaddr + 8)};
```

lioncash commented on tests/A64/testenv.h:
```diff
@@ -125,3 +125,89 @@ class A64TestEnv final : public Dynarmic::A64::UserCallbacks {
+    u64 ticks_left = 0;
+    char* backing_memory = nullptr;
+
+    RawMemoryTestEnv(char* addr) : backing_memory(addr) {}
```
```suggestion
    explicit RawMemoryTestEnv(char* addr) : backing_memory(addr) {}
```

lioncash commented on tests/A64/testenv.h:
```diff
@@ -125,3 +125,89 @@ class A64TestEnv final : public Dynarmic::A64::UserCallbacks {
+    }
+
+    std::uint8_t MemoryRead8(u64 vaddr) override {
+        return *(backing_memory + vaddr);
```
```suggestion
        return static_cast<std::uint8_t>(backing_memory[vaddr]);
```

lioncash commented on tests/A64/testenv.h:
```diff
@@ -125,3 +125,89 @@ class A64TestEnv final : public Dynarmic::A64::UserCallbacks {
+    RawMemoryTestEnv(char* addr) : backing_memory(addr) {}
+
+    std::uint32_t MemoryReadCode(u64 vaddr) override {
+        return *((uint32_t*)backing_memory + (vaddr / 4));
```
This should be using memcpy to extract the value from the memory, otherwise this is undefined behavior.

Ditto for other occurrences.

lioncash replied:
This needs to be using the read helper as well.

lioncash commented on tests/A64/testenv.h:
```diff
@@ -125,3 +125,89 @@ class A64TestEnv final : public Dynarmic::A64::UserCallbacks {
+        backing_memory[vaddr] = value;
+    }
+    void MemoryWrite16(u64 vaddr, std::uint16_t value) override {
+        *((uint16_t*)backing_memory + (vaddr / 2)) = value;
```
Ditto for setting as well.

MerryMage commented:
Please match the codestyle of the rest of the codebase.

Design comments:

1. Handling of segmentation faults is required. Users of this library intentionally map inaccessible memory into a fastmem region to trigger this behaviour. This must fallback to another method of memory access. Recompilation should be a configuration option.
2. Not all users of this library find mirroring the address space desirable behaviour. This is why it’s a configuration option.
3. You can reserve a register for the fastmem base pointer.

Misc comments:

1. Your memory read/write implementations in `RawMemoryTestEnv` are incorrect. Misaligned memory accesses need to be supported.

The above is from a brief look. I will give this a proper review when I can.

MerryMage commented on include/dynarmic/A64/config.h:
```diff
@@ -171,7 +171,7 @@ struct UserConfig {
     /// Declares how many valid address bits are there in virtual addresses.
     /// Determines the size of page_table. Valid values are between 12 and 64 inclusive.
-    /// This is only used if page_table is not nullptr.
+    /// This is only used if page_table or fastmem_pointer is not nullptr.
```
Incorrect documentation change.

vdwjeremy replied:
actually, page_table_address_space_bits is reused for fastmem as well, as the semantic was quite similar. However if this is misleading I can introduce a new parameter.

MerryMage commented on include/dynarmic/A64/config.h:
```diff
@@ -196,6 +196,11 @@ struct UserConfig {
     bool only_detect_misalignment_via_page_table_on_page_boundary = false;
 
+    // Fastmem Pointer
+    // This should point to the beginning of a 128 bits aligned 2^page_table_address_space_bits address space which is in arranged just like
```
The fastmem pointer doesn't have to be 128 bits aligned.

vdwjeremy replied:
true, although it seems to be recommended : https://software.intel.com/en-us/forums/intel-isa-extensions/topic/752392

lioncash commented on tests/A64/testenv.h:
```diff
@@ -125,3 +125,96 @@ class A64TestEnv final : public Dynarmic::A64::UserCallbacks {
+        return *((uint32_t*)backing_memory + (vaddr / 4));
+    }
+
+    template<typename rtype> rtype read(u64 vaddr) {
```
```suggestion
    template<typename T>
    T Read(u64 vaddr) {
```

vdwjeremy commented:
> Apologies for my stricter than usual comments, but this touches a part of the codebase which is difficult to adequately test.

no worry, that's helpful :)

vdwjeremy commented:
> Design comments:
>     1. Handling of segmentation faults is required. Users of this library intentionally map inaccessible memory into a fastmem region to trigger this behaviour. This must fallback to another method of memory access. Recompilation should be a configuration option.
>     2. Not all users of this library find mirroring the address space desirable behaviour. This is why it’s a configuration option.
>     3. You can reserve a register for the fastmem base pointer.

1. ok, I was not sure this could be useful, in that case I will add this part
2. can you develop ?
3. will do, the register will be reserved for the full block though, do you foresee potential issues ?

MerryMage commented:
Just FYI, I've implemented what I've said above here: https://github.com/MerryMage/dynarmic/compare/fm

vdwjeremy commented:
> Just FYI, I've implemented what I've said above here: https://github.com/MerryMage/dynarmic/compare/fm

Too fast !
I was working on a similar, but it won't be compatible I'm afraid, and I didn't have the R13 register yet. Do you want to continue on branch fm ?

On the exception fallback I think we had the same, but I couldn't validate it succesfully, it goes through ```A64EmitX64::FastmemCallback``` but then crashes.
I'm testing with restrictions on windows pages such as
```
    void* vmem = VirtualAlloc(nullptr, memory_size, MEM_RESERVE, PAGE_NOACCESS);
    VirtualAlloc(vmem, 1024, MEM_COMMIT, PAGE_READWRITE); // code memory
    VirtualAlloc((char*)vmem+ 1024 * 1024, 1024 * 1024, MEM_COMMIT, PAGE_NOACCESS); // no access
```
Do you get better results ?

vdwjeremy commented:
rebased the PR on top of fm branch, just changed :
- added unit tests
- fixed conditions to use user callbacks
The few tests I did on fallback still can't clear the SIGSEGV when faulting on memory though

MerryMage commented:
Hey! Apologies for getting back to this late. LGTM as is, I'll rebase onto `mp` when I can.

MerryMage commented:
> can't clear the SIGSEGV

Could you clarify what you mean by this?

vdwjeremy commented:
OK I understood what happens.
Effectivelly the MSVC debugger intercepts the segfault, that can be resumed properly and does not appear when the debugger is not attached.
But also the test framework interferes with signals. I could get a different behaviour by playing with these flags
```
CATCH_CONFIG_NO_POSIX_SIGNALS
CATCH_INTERNAL_CONFIG_NO_POSIX_SIGNALS
CATCH_CONFIG_NO_WINDOWS_SEH
CATCH_INTERNAL_CONFIG_NO_WINDOWS_SEH
```
But it starts to be messy and I can't find a way to make a portable test out of it. For now I will leave it as is, if that's ok with you.

degasus commented:
Is there any way how I can help to get this merged to master as well? I can confirm that this implementation is working fine. I've seen a 15% performance increase in the CPU emulation in yuzu ingame BOTW.

MerryMage commented:
I can rebase this branch to a commit which is yuzu-compatible. (I've removed some yuzu-specific hacks.)

degasus commented:
Thanks for the offer, I already did so: https://github.com/degasus/dynarmic/commits/fm
That's how I've tested this branch. However I think in the long term, this should all go to master, shouldn't it?
